crates/hir-def/src/body/lower.rs

```diff
@@ -302,16 +302,33 @@ impl ExprCollector<'_> {
                 self.alloc_expr(Expr::For { iterable, pat, body, label }, syntax_ptr)
             }
             ast::Expr::CallExpr(e) => {
-                let callee = self.collect_expr_opt(e.expr());
-                let args = if let Some(arg_list) = e.arg_list() {
-                    arg_list.args().filter_map(|e| self.maybe_collect_expr(e)).collect()
-                } else {
-                    Box::default()
+                let is_rustc_box = {
+                    let attrs = e.attrs();
+                    attrs.filter_map(|x| x.as_simple_atom()).any(|x| x == "rustc_box")
                 };
-                self.alloc_expr(
-                    Expr::Call { callee, args, is_assignee_expr: self.is_lowering_assignee_expr },
-                    syntax_ptr,
-                )
+                if is_rustc_box {
+                    if let Some(arg) = e.arg_list().into_iter().flat_map(|x| x.args()).next() {
+                        let expr = self.collect_expr(arg);
+                        self.alloc_expr(Expr::Box { expr }, syntax_ptr)
+                    } else {
+                        self.alloc_expr(Expr::Missing, syntax_ptr)
+                    }
```

Comment:
Shouldn't this be 
```suggestion
                        let expr = self.collect_expr_opt(e.arg_list().into_iter().flat_map(|x| x.args()).next());
                        self.alloc_expr(Expr::Box { expr }, syntax_ptr)
```
?